prov/tcp/src/tcpx_ep.c

```diff
@@ -209,18 +260,39 @@ static int tcpx_pep_sock_create(struct tcpx_pep *pep)
 		return -FI_EIO;
 	}
 
-	ret = tcpx_setup_socket(pep->sock);
-	if (ret) {
-		goto err;
-	}
+	if (ofi_addr_get_port(pep->info->src_addr) != 0) {
+		ret = tcpx_setup_socket(pep->sock);
+		if (ret) {
+			goto err;
+		}
+		ret = bind(pep->sock, pep->info->src_addr,
+		      (socklen_t) pep->info->src_addrlen);
+
+		if (ret) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to bind listener: %s\n",
+				strerror(ofi_sockerr()));
+			goto err;
+		}
+	} else {
 
-	ret = bind(pep->sock, pep->info->src_addr,
-		   (socklen_t) pep->info->src_addrlen);
-	if (ret) {
-		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
-			"failed to bind listener: %s\n",
-			strerror(ofi_sockerr()));
-		goto err;
+		if ((pep->port_range.low  == 0) &&
+		    (pep->port_range.high == 0)) {
+			ret = tcpx_setup_socket(pep->sock);
+			if (ret) {
+				goto err;
+			}
+		} else {
+			ret = tcpx_setup_socket_nodelay(pep->sock);
+			if (ret) {
+				goto err;
+			}
+		}
```

Comment:
Why isn't tcpx_setup_socket() just called in all cases?

Reply:
@shefty Since we select the port automatically at the provider level, we have to go through the range of ports and find the first free one.
In the case of `SO_REUSEADDR`, `bind ()` allows the port to be reused, which breaks the logic of port selection and as the results, it's a deadlock.